### PR TITLE
Story/quad 137

### PR DESCRIPTION
**Pull request by PawanMahalle**

### CODE REVIEW

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public String getIdPrefix() {
+        return messages.getProperty("projectblog_id.prefix");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ProjectBlogEntryDTO getProjectBlogDTO(String id) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java]: getProjectBlogEntryDTO

**Comment by jdamerow** [comment]
ok to test

**Comment by jdamerow** [comment]
The "Add project blog entry" button should only be visible if the user is logged in and has the rights to add a new blog entry.
image

File: Quadriga/src/main/webapp/WEB-INF/views/addprojectblogentry.jsp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+				toolbar : 'undo redo | styleselect | bold italic | alignleft aligncenter alignright alignjustify | bullist numlist outdent indent | link image | print preview media fullpage | forecolor backcolor emoticons'
+			});
+</script>
+
+<article class="is-page-content">
+	<form:form commandName="projectBlogEntry" method="POST"
+		action="${pageContext.servletContext.contextPath}/sites/${project.projectId}/addprojectblogentry">
+
+		<header>
+			<h2>Create new Project Entry</h2>
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/WEB-INF/views/addprojectblogentry.jsp]: Should be "Create new blog entry"

File: Quadriga/src/main/webapp/WEB-INF/views/addprojectblogentry.jsp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+			<span class="byline">Please fill in blog details</span>
```
Comment (jdamerow) [inline Quadriga/src/main/webapp/WEB-INF/views/addprojectblogentry.jsp]: delete this line

**Comment by jdamerow** [comment]
The button should be aligned with the other blue button (on the right side) and should have a little space from the input box:
image

**Comment by jdamerow** [comment]
I'm getting a null pointer exception when I try to add an entry:

```
java.lang.NullPointerException
    edu.asu.spring.quadriga.mapper.ProjectBlogEntryDTOMapper.getProjectBlogEntryDTO(ProjectBlogEntryDTOMapper.java:46)
    edu.asu.spring.quadriga.service.impl.projectblog.ProjectBlogEntryManager.addNewProjectBlogEntry(ProjectBlogEntryManager.java:44)
    sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke(Method.java:497)
    org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:302)
    org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:190)
    org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
    org.springframework.transaction.interceptor.TransactionInterceptor$1.proceedWithInvocation(TransactionInterceptor.java:99)
    org.springframework.transaction.interceptor.TransactionAspectSupport.invokeWithinTransaction(TransactionAspectSupport.java:281)
    org.springframework.transaction.interceptor.TransactionInterceptor.invoke(TransactionInterceptor.java:96)
    org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
    org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:208)
    com.sun.proxy.$Proxy1541.addNewProjectBlogEntry(Unknown Source)
    edu.asu.spring.quadriga.web.projectblog.AddProjectBlogEntryController.addProjectBlogEntryRequest(AddProjectBlogEntryController.java:111)
    edu.asu.spring.quadriga.web.projectblog.AddProjectBlogEntryController$$FastClassBySpringCGLIB$$ce11c6ef.invoke(<generated>)
    org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
    org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:720)
    org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
    org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:85)
    edu.asu.spring.quadriga.aspects.AccessAspect.chkAuthorization(AccessAspect.java:83)
    sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke(Method.java:497)
    org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethodWithGivenArgs(AbstractAspectJAdvice.java:620)
    org.springframework.aop.aspectj.AbstractAspectJAdvice.invokeAdviceMethod(AbstractAspectJAdvice.java:609)
    org.springframework.aop.aspectj.AspectJAroundAdvice.invoke(AspectJAroundAdvice.java:68)
    org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:168)
    org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint.proceed(MethodInvocationProceedingJoinPoint.java:97)
    edu.asu.spring.quadriga.aspects.InjectProjectAspect.injectProject(InjectProjectAspect.java:104)
    sun.reflect.GeneratedMethodAccessor669.invoke(Unknown Source)
    sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
    java.lang.reflect.Method.invoke(Method.java:497)
    ...
```

File: Quadriga/src/main/resources/locale/projectconstants.properties [hunk 4]
```diff
@@ -6,4 +6,5 @@ dictionary_id.prefix=DICT
 concept_collection_id.prefix=CC
 workspace_id.prefix=WS
 network_id.prefix=NET
-notype_id.prefix=ID
\ No newline at end of file
+notype_id.prefix=ID
+projectblogentry_id.prefix=PROJBLOGENTRY
```
Comment (jdamerow) [inline Quadriga/src/main/resources/locale/projectconstants.properties]: I think I mentioned this a couple of times now. This should be only 3 or 4 characters long, e.g. BLGE.

**Comment by PawanMahalle** [comment]
story/QUAD-137
1. Implemented review comments

**Comment by PawanMahalle** [comment]
Currently working on form validation.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    @Override
+    public String getIdPrefix() {
+        return messages.getProperty("projectblogentry_id.prefix");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public ProjectBlogEntryDTO getProjectBlogEntryDTO(String id) {
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java]: Is there a reason you have two methods that do the same? This one and `getDTO`?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    @Override
+    @Transactional
+    public List<ProjectBlogEntryDTO> getProjectBlogEntryDTOList(String projectId) throws QuadrigaStorageException {
+       
+        if (projectId == null || projectId.equals(""))
+            return null;
+
+        // Create a query to get all projects
+        Query query = sessionFactory.getCurrentSession().getNamedQuery(
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java]: Let's catch the HibernateException here and throw an appropriate QuadrigaStorageException.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                "ProjectBlogEntryDTO.findByProjectId");
+        query.setParameter("projectId", projectId);
+
+        @SuppressWarnings("unchecked")
+        List<ProjectBlogEntryDTO> projectBlogEntryDTOList = query.list();
+
+        return projectBlogEntryDTOList;
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java:71]: merge with line 69

File: Quadriga/src/main/java/edu/asu/spring/quadriga/mapper/ProjectBlogEntryDTOMapper.java [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+ * instance to {@linkplain ProjectBlogEntryDTO} instance.
+ * 
+ * @author PawanMahalle
+ *
+ */
+@Service
+public class ProjectBlogEntryDTOMapper extends BaseMapper {
+
+    @Autowired 
+    IUserManager userManager;
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/mapper/ProjectBlogEntryDTOMapper.java]: should be private

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+     *            Model instance
+     * @return view to add new project blog entry
+     * @throws QuadrigaStorageException
+     * @throws QuadrigaAccessException
+     *             This exception is thrown when user other than project admin
+     *             and owner tries to add project blog entry.
+     */
+    @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 1, userRole = {
+            RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
+    @RequestMapping(value = "sites/{projectId}/addprojectblogentry", method = RequestMethod.GET)
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java]: You don't have the project id here. You have the unix name. Didn't I comment on this the last time?

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+     * @param projectId
+     * @param model
+     * @param principal
+     * @return
+     * @throws QuadrigaStorageException
+     * @throws QuadrigaAccessException
+     */
+    @AccessPolicies({ @ElementAccessPolicy(type = CheckedElementType.PROJECT, paramIndex = 3, userRole = {
+            RoleNames.ROLE_COLLABORATOR_ADMIN, RoleNames.ROLE_PROJ_COLLABORATOR_ADMIN }) })
+    @RequestMapping(value = "sites/{projectId}/addprojectblogentry", method = RequestMethod.POST)
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java]: same here

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+     * @throws NoSuchRoleException
+     */
+    @RequestMapping(value = "sites/{ProjectUnixName}/projectblog", method = RequestMethod.GET)
+    public String projectBlog(@PathVariable("ProjectUnixName") String projectUnixName, Model model,
+            @InjectProject(unixNameParameter = "ProjectUnixName") IProject project, Principal principal)
+            throws QuadrigaStorageException, NoSuchRoleException {
+
+        // Obtain project blog entries for given project
+        String projectId = project.getProjectId();
+        List<IProjectBlogEntry> projectBlogEntryList = projectBlogEntryManager.getProjectBlogEntryList(projectId);
```
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java]: You probably don't want to get all blog entries here or the page would be enormous at some point. Let's start with just the last n entries (configurable by the admin).

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java (hunk 5 above)
Comment (PawanMahalle) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java]: The getProjectBlogEntryDTO() acts as wrapper to generic method getDTO(). The getDTO() needs type of class as method parameter while getProjectBlogEntryDTO() asks for just the id as parameter.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java (hunk 7 above)
Comment (PawanMahalle) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java:71]: Merged the lines.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java (hunk 9 above)
Comment (PawanMahalle) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/projectblog/AddProjectBlogEntryController.java]: Passed projectId as request parameter.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java (hunk 11 above)
Comment (PawanMahalle) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/web/publicwebsite/PublicBlogController.java]: obtained the number of blog entries to fetch from user.properties file

**Comment by jdamerow** [comment]
please resolve conflicts and read git documentation about auto crlf. Make sure you don't submit files with windows line ending.

File: Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java (hunk 5 above)
Comment (jdamerow) [inline Quadriga/src/main/java/edu/asu/spring/quadriga/dao/impl/projectblog/ProjectBlogEntryDAO.java]: That's why you have to implement below `getDTO(String id)` method. Remove this one.

**Comment by kavinya** [comment]
@jdamerow - story/QUAD-137 : Resolved Merge Conflicts

**Comment by PawanMahalle** [comment]
Created a new [pull request](https://github.com/diging/quadriga/pull/221) to avoid unnecessary files in this PR.

**Pull request closed without merging**